Actual Report for Reference Setup, Performance, Scalability, and Sizing Guidelines (#7905)
longhorn/longhorn: scalability/reference-setup-performance-scalability-and-sizing-guidelines/public-cloud/medium-node-spec.md

**Comments:**
* We observe Longhorn volume achieves much bigger sequential read IOPs compares to the local path volume. 
  This can be explained by 2 factors:
  * Each Longhorn volume has 3 replicas so it can read concurrently from multiple replicas (each replica is on a different EBS disk) while the local path volume is reading from single EBS disk
  * When running Kbench with Longhorn volume, there is IO merging at both the kernel layer and EBS layer. 
    The IO merging can cause multiple Kbench IOs to be submitted as 1 therefore increasing the IOPs.
    This IO merging effect doesn't happen when running Kbench with local path volume.
    See more detailed explanation at https://github.com/longhorn/longhorn/pull/7905#issuecomment-1945463272

Next, we use Kbench with Longhorn PVCs and scale up the number of Kbench pods to see how sequential read IOPs is effected when there are more and more IO intensive pods.

Scaling workload from 1 to 3 pods.

> Result:
> * Each Kbench pod is able to achieve 19840 sequential read IOPs on its Longhorn volume

> > * Total random IOPs can be achieved by all 3 Longhorn volumes is: 59520

Reviewer: This  is the one I was aiming for with those other comments. Not sure how I got it mixed up.

```suggestion
> * Total sequential IOPs can be achieved by all 3 Longhorn volumes is: 59520
```

Author: Fixed. Thank you

**Comment:**
* Since each EBS volume on the host is provisioned with 360 MiB/s. Total IOPs of 3 EBS volumes is around 1080 MiB/s
It looks like Longhorn system is able to reach the maximum IOPs capacity of the 3 host EBS volumes. 
Longhorn is able to achieve slightly above 1080 MiB/s maybe due to small number of IO mergings or EBS actual bandwidth might be a bit higher the specified limit. 

Scaling workload from 3 to 6, then 6 to 9, then 9 to 12, then 12 to 15

> Result:
> * At 6 pods, the average random read bandwidth per Longhorn volume is 196 MiB/s. Total random bandwidth is 1176 MiB/s
> * At 9 pods, the average random read bandwidth per Longhorn volume is 131 MiB/s. Total random bandwidth is 1179 MiB/s
> * At 12 pods, the average random read bandwidth per Longhorn volume is 97.5 MiB/s. Total random bandwidth is 1170 MiB/s
> * At 15 pods, the average random read bandwidth per Longhorn volume is 77.5 MiB/s. Total random bandwidth is 1162 MiB/s

<img src="./assets/medium-node-spec/read-performance/random-read-bandwidth-6-9-12-15.png" alt="drawing" style="width:600px;"/>

**Comment:**

> * From the scaling test so far, we can see that the total random read bandwidth of all Longhorn volumes remain relative same around 1160 MiB/s when the number of Kbench pods increase. 

Author: Thank you. Fixed

        resources:
          requests:
            storage: 20Gi
```
</details>

Some of the important Kbench parameters we would like to call out are:
* `MODE: sequential-read-bandwidth`: specifies that we are running sequential read bandwidth job # TODO add the link to the job here
* `SIZE: 15G`: the `fio` test size is 15G, this will avoid cache effect
* PVC size is 20G

> Result:
> * Local Path Provisioner: 362 MiB/s
> * Longhorn: 669 MiB/s

<img src="./assets/medium-node-spec/read-performance/sequential-read-bandwidth-local-path-vs-longhorn.png" alt="drawing" style="width:600px;"/>

**Comment**:
* Because Longhorn has 3 replicas, it can read from 3 replicas concurrently thus may produce better read performance
* Note that the sequential read bandwidth is smaller than the random read bandwidth when testing against a single Longhorn volume of 3 replicas.
  We observed that this behavior happen when there is IO merging happen at kernel level and when the Longhorn volume has multiple replicas on different nodes.
  We don't understand the root cause of this behavior yet so we created a GitHub ticket to investigate it later https://github.com/longhorn/longhorn/issues/8108 

> Next, we use Kbench with Longhorn PVCs and scale up the number of Kbench pods to see how sequential read bandwidth is effected when there are more and more IO intensive pods.

Author: Thank you. Fixed

          readinessProbe:
            exec:
              command:
                - sh
                - -c
                - '[ "$(ls -A /test-result)" ]'
            initialDelaySeconds: 30
            periodSeconds: 10
          command:
            - metric-exporter
            - -d
            - start
          env:
            - name: DATA_DIR
              value: /test-result
            - name: VOLUME_ACCESS_MODE
              value: rwo
            - name: TEST_MODE
              value: read-only
            - name: RATE_LIMIT_TYPE
              value: no-rate-limit
          ports:
            - containerPort: 8080
              name: metrics
          volumeMounts:
            - name: vol
              mountPath: /volume/
            - name: shared-data
              mountPath: /test-result
      volumes:
        - name: shared-data
          emptyDir: {}
  volumeClaimTemplates:
    - metadata:
        name: vol
      spec:
        accessModes: ["ReadWriteOnce"]
        storageClassName: "longhorn"
        resources:
          requests:
            storage: 20Gi
```
</details>

Some of the important Kbench parameters we would like to call out are:

> * `MODE: random-read-latency`: specifies that we are running random read latency job # TODO add the link to the job here

Author: Thank you. I fixed the TODO comments. However, I think no need to add more summary since the link to the job is already clear. WDYT?

* `SIZE: 15G`: the `fio` test size is 15G, this will avoid cache effect
* PVC size is 20G

> Result:
> * Local Path Provisioner: 500 microsecond
> * Longhorn: 750 microsecond

<img src="./assets/medium-node-spec/read-performance/random-read-latency-local-path-vs-longhorn.png" alt="drawing" style="width:600px;"/>

**Comment**:

> * Because IO path of Longhorn volume is longer than local path provisioner, it is expected that Longhorn volume has bigger random read latency 

Author: Fixed. Thank you

> Result:

| Number of Kbench pods | Average Random Read Latency (ns) | 
|-----------------------|----------------------------------|
| 1                     | 742955                           |
| 3                     | 750170                           | 
| 6                     | 774555                           | 
| 9                     | 790714                           |
| 12                    | 810890                           |
| 15                    | 832421                           |
| 18                    | 858495                           |
| 21                    | 876277                           |
| 24                    | 902481                           |
| 27                    | 934727                           |
| 30                    | 972684                           |
| 33                    | 1014888                          |
| 36                    | 1117489                          |
| 39                    | 1212721                          |
| 42                    | 1304671                          |
| 45                    | 1391987                          |
| 48                    | 1488104                          |

> | 51                    | 1581291                          |

Author: Unlike the IOPs or Bandwidth which have stable numbers for the test duration. These latency tests have moving numbers so I think I would like to record exact data points so we can compare when we have the new engine's front end (which reduced the latency). WDYT?

> | 51                    | 1581291                          |

Reviewer: I'm fine with it. Thanks for the explanation.

            - name: shared-data
              mountPath: /test-result
      volumes:
        - name: shared-data
          emptyDir: {}
  volumeClaimTemplates:
    - metadata:
        name: vol
      spec:
        accessModes: ["ReadWriteOnce"]
        storageClassName: "longhorn"
        resources:
          requests:
            storage: 20Gi
```
</details>

Some of the important Kbench parameters we would like to call out are:
* `MODE: random-write-iops`: specifies that we are running random write iops job # TODO add the link to the job here
* `SIZE: 15G`: the `fio` test size is 15G, this will avoid cache effect
* PVC size is 20G

> Result:
> * Local Path Provisioner: 9980
> * Longhorn: 1150

<img src="assets/medium-node-spec/write-performance/random-write-iops-local-path-vs-longhorn.png" alt="drawing" style="width:600px;"/>

**Comment**:
* Longhorn write speed is slower than local path volume because of several factors:
  * Because Longhorn has 3 replicas, it has to write to 3 replicas for each IO
  * By default, Longhorn has revision counter enabled which means that for each write IO, Longhorn has to do an additional write IO to write the revision counter

>   * Longhorn v1 tgt-iscsi stack introduces some overhead

Author: Thank you. Fixed

Next, we use Kbench with Longhorn PVCs and scale up the number of Kbench pods to see how random write IOPs is effected when there are more and more IO intensive pods.

Scaling workload from 1 to 3 pods.
> Result:
> * Each Kbench pod is able to achieve 1180 random write IOPs on its Longhorn volume
> * Total random write IOPs can be achieved by all 3 Longhorn volumes is 3540

> <img src="assets/medium-node-spec/write-performance/random-write-iops-1-3.png" alt="drawing" style="width:600px;"/>

Reviewer: This is one of those graphs that would definitely benefit from y-axis scaling (as @james-munson mentioned previously). My first thought when I looked at it was, "Wow, why do the write IOPs vary so much?!" when they really don't.

I'm guessing you can't easily obtain a scaled version of this graph at this point. If this is the case, lets ignore it for this version of the report.

Reviewer: There are other similar graphs. I will not make this comment on those as well.

Author: Thank you for pointing this out. Yes, I am aware of this issue. I fixed the graph for the new tests but the completed tests we don't have fixed graphs. We would have to rerun the test for the new graphs. Let's have this improvement in the next report

> * Longhorn: 159 MiB/s

<img src="./assets/medium-node-spec/write-performance/random-write-bandwidth-local-path-vs-longhorn.png" alt="drawing" style="width:600px;"/>

**Comment**:
* Longhorn write speed is slower than local path volume because of several factors:
    * Because Longhorn has 3 replicas, it has to write to 3 replicas for each IO
    * By default, Longhorn has revision counter enabled which means that for each write IO, Longhorn has to do an additional write IO to write the revision counter
    * Longhorn v1 tgt-iscsi stack introduces some overhead

Next, we use Kbench with Longhorn PVCs and scale up the number of Kbench pods to see how random write bandwidth is effected when there are more and more IO intensive pods.

Scaling workload from 1 to 3 pods.

> Result:
> * Each Kbench pod is able to achieve 118 MiB/s random write bandwidth on its Longhorn volume
> * Total random write bandwidth can be achieved by all 3 Longhorn volumes is 354 MiB/s

<img src="./assets/medium-node-spec/write-performance/random-write-bandwidth-1-3.png" alt="drawing" style="width:600px;"/>

**Comment:**
* Since each EBS volume on the host is provisioned with 360 MiB/s. 

>   It looks like Longhorn system is able to reach the maximum bandwidth capacity of the host EBS disks

Reviewer: NIT: I can't remember if this is explained before. If it is, feel free to ignore.

IIUC, this statement is true because each Longhorn volume writes to 3 replicas and 354 * 3 is close to 360 * 3? If so, maybe we need to spell it out more?

Author: >IIUC, this statement is true because each Longhorn volume writes to 3 replicas and 354 * 3 is close to 360 * 3? If so, maybe we need to spell it out more?

You are correct. A simpler explanation is: a Longhorn volume can never write faster than its slowest replica. So, Total write of all Longhorn volumes should be capped at speed of the slowest EBS host.

| 12                    | 2253089                           |
| 15                    | 2807773                           |
| 18                    | 3373824                           |
| 21                    | 3938172                           |
| 24                    | 4503630                           |
| 27                    | 5071816                           |
| 30                    | 5640033                           |
| 33                    | 6208153                           |
| 36                    | 6771033                           |
| 39                    | 7341914                           |
| 42                    | 7905373                           |
| 45                    | 8480925                           |
| 48                    | 9040640                           |
| 51                    | 9609189                           |

<img src="./assets/medium-node-spec/write-performance/random-write-latency-1-51.png" alt="drawing" style="width:600px;"/>

**Comment**:
* As the number of kbench pods increase:
  * From 0 to 12 pods, the average random write latency of each Longhorn volume increase in linear fashion with smaller rate

>   * From 12 to 51 pods, the average random write latency of each Longhorn volume increase in linear fashion with bigger rate

Reviewer: I think I agree with your assessment, but curious what your criteria was for determining the graph is linear? Eyeballing it? Or did you apply some sort of mathematical test?

Author: It was Eyeballing for this report. We can draw a derivative graph to confirm if needed in the future. WDYT?

Reviewer: I think it is fine. Just curious mostly.

| 27                    | 5078192                               |
| 30                    | 5640632                               |
| 33                    | 6204730                               |
| 36                    | 6770942                               |
| 39                    | 7343912                               |
| 42                    | 7906260                               |
| 45                    | 8475788                               |
| 48                    | 9035704                               |
| 51                    | 9611938                               |

<img src="./assets/medium-node-spec/write-performance/sequential-write-latency-1-51.png" alt="drawing" style="width:600px;"/>

**Comment**:
* As the number of kbench pods increase:
    * From 0 to 12 pods, the average sequential write latency of each Longhorn volume increase in linear fashion with smaller rate
    * From 12 to 51 pods, the average sequential write latency of each Longhorn volume increase in linear fashion with bigger rate

## Longhorn Control Plane Performance

We deploy a StatefulSet workload. 
Each pod of the StatefulSet is a nginx web server with 1 Longhorn PVC attached to it.

> Each pods as liveness probe to make sure the Longhorn volume is still functional.

Reviewer: ```suggestion
Each pods has a liveness probe to ensure the Longhorn volume is still functional.
```

Author: Fixed. Thank you

<summary>StatefulSet workload</summary>
<br>

```yaml
apiVersion: v1
kind: Service
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  ports:
  - port: 80
    name: web
  selector:
    app: nginx
  type: NodePort
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: web
  namespace: default
spec:
  podManagementPolicy: Parallel
  selector:
    matchLabels:
      app: nginx
  serviceName: "nginx"
  replicas: 0
  template:
    metadata:
      labels:
        app: nginx
    spec:
      restartPolicy: Always
      terminationGracePeriodSeconds: 10
      containers:
      - name: nginx
        image: registry.k8s.io/nginx-slim:0.8
        livenessProbe:
          exec:
            command:
              - ls
              - /usr/share/nginx/html/lost+found
          initialDelaySeconds: 5
          periodSeconds: 5
        ports:
        - containerPort: 80
          name: web
        volumeMounts:
        - name: www
          mountPath: /usr/share/nginx/html
  volumeClaimTemplates:
  - metadata:
      name: www
    spec:
      accessModes: [ "ReadWriteOnce" ]
      storageClassName: "longhorn"
      resources:
        requests:
          storage: 1Gi
```

</details>

> Then we scale up the StatefulSet with the rate 30 pods per minutes and monitor the number of pods, CPU and RAM usage of Longhorn system, ECTD, and KubeAPI metrics.

Reviewer: Do you think we should link to the mechanism we used to do the scaling from this report?

Author: Sure. Added. Thank you

We define the following stopping conditions for the scale test. 
If we hit any of these conditions, we stop the scale test and record the current scaling number.
1. There are more than 10 pods with starting time more than 4 minutes (the duration from pod being created to pod becomes running). 

>    The idea is that if pod is taking more than 4 minutes to start, the system is being too slow

Reviewer: We discussed this condition in the team sync and felt it was a bit weaker than we wanted because:

- We measure it cumulatively. Meaning at any given moment, we may only have one pod taking a long time to start even though there have been ten such pods over the life of the test.
- If, at this moment, there is only one such pod, Longhorn seems relatively stable.

However, looking at the graph, I think condition is good enough. At the moment it is broken, we see drastic increase from zero such pods to many. We should probably just look carefully at the graph each time we run a similar test in the future to make sure it looks similar.

1. The per-minute rate of pod become running is smaller than 10. It means that the system should be able to create more than 10 pods per minute. 
   If not, it is being too slow
1. There are more than 3 crashing pods or restarted containers. If there are many crashes, the system is not stable at the scale

Finally, we will scale down the StatefulSet from the max value to 0 with the rate of 30 pods per minutes. We will measure the scaling down speed.

This test design touches the end-to-end life cycle of a pod from pod creation -> CSI flow provision/attach/mount Longhorn volume -> CSI flow unmount/detach Longhorn volume -> pod termination

> Notes: There are 2 small modification we made in this section comparing to the previous data plan performance testing section:
> 1. Because of this known issue in Longhorn v1.6.0 https://github.com/longhorn/longhorn/issues/7919, it might lock up the Longhorn control plane when there are many volumes.
>    Therefore, we decided to use Longhorn v1.6.1 version for this control plane test 
>    Longhorn v1.6.1 should have very similar characteristic as Longhorn 1.6.0 except for some bug fixes

> > 1. At the big scale of workload, Prometheus server is eating too much RAM and CPU. 

Reviewer: ```suggestion
> 1. At this large scale, Prometheus server uses too much RAM and CPU. 
```

Author: FIxed. Thank you

1. We can see that the number of running pod is linearly scaled from 0 to 840. Then rate is slow down after that
1. The component which is consuming the most CPU and RAM are instance manager pods followed by longhorn manager pods
1. API and ETCD is not overloaded

#### Scaling Down

##### Comment

Scaling down has no issue. Everything is perfectly linear similar to the above test

### 1 master node + 9 worker nodes

#### Result
1. The cluster can handle up to 1561 pods with 1561 Longhorn volumes
1. Each node can have 173 Longhorn volumes
2. Each node can have 520 Longhorn replicas

#### Scaling Up

<img src="./assets/medium-node-spec/control-plane-performance/1-master-9-workers-scale-up.png" alt="drawing"/>

##### Comment
1. This test hits the 1st stopping condition (there are more than 10 pods with starting time more than 4 minutes) at 1561 pods (with 1561 Longhorn volumes)
1. We can see that the number of running pod is linearly scaled from 0 to 1400. Then rate is slow down after that
1. The component which is consuming the most CPU and RAM are instance manager pods followed by longhorn manager pods

> 1. API and ETCD is not overloaded

Reviewer: ```suggestion
1. API and ETCD are not overloaded
```

Author: Thank you. Fixed

#### Scaling Down

##### Comment

Scaling down has no issue. Everything is perfectly linear similar to the above test

> ## Volume Maximum Actual Size

Reviewer: I think we are still waiting on an improvement for this part based on internal discussions.

Reviewer: This is all that's needed for an approval now IMO.

See more at https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-optimized.html .

### Network Spec
* Network bandwidth: 
  * Baseline bandwidth 10 Gbps
  * Burst bandwidth: Up to 25 Gbps in short time 
* Network latency: ~0.3ms RTT via ping command

### Disk Spec
* We are using dedicated disk for Longhorn volumes' replicas on the nodes. 
We select a typical EBS volume with average IOPs and bandwidth performance:
  * Single EC2 EBS gp3
  * 1TB 
  * IOPs set to 10000 
  * Throughput set to 360MiB/s 
  * Formatted as ext4 filesystem

**Comment:**
* We choose 10000 for EBS disk's IOPs simply because it is a middle value between minimum value 3000 and maximum value 16000 of the gp3 EBS disk
* We choose 360MiB/s for EBS disk's bandwidth because the m5zn.2xlarge EC2 instance has EBS bandwidth 396.25 MiB/s. 
  If we choose a bigger value than 396.25 MiB/s for EBS disk's bandwidth, the ec2 instance would not be able to push EBS disk to that value

### Kubernetes Spec
* Kubernetes Version: v1.27.10+rke2r1
* CNI plugin: Calico
* Control plane nodes are separated from worker nodes

### Longhorn Config
* Longhorn version: v1.6.0

> * Settings:

Author: Good point! However, in the test we are running there isn't anything else consume the network other than Longhorn (control plane doesn't seem to use much networking I think) so I think it might have less value to use dedicated network since it is already dedicated in some sense? WDYT?

> * Settings:

Reviewer 2: Oh, I totally agree with the decision, we just need to note it in the description of the config.

          ports:
            - containerPort: 8080
              name: metrics
          volumeMounts:
            - name: vol
              mountPath: /volume/
            - name: shared-data
              mountPath: /test-result
      volumes:
        - name: shared-data
          emptyDir: {}
  volumeClaimTemplates:
    - metadata:
        name: vol
      spec:
        accessModes: ["ReadWriteOnce"]
        storageClassName: "longhorn"
        resources:
          requests:
            storage: 20Gi
```
</details>

Some of the important Kbench parameters we would like to call out are:
* `MODE: random-read-iops`: specifies that we are running [random read iops job](https://github.com/longhorn/kbench/blob/main/fio/bandwidth-random-read.fio)
* `SIZE: 15G`: the `fio` test size is 15G, this will avoid cache effect
* PVC size is 20G

> Result: 
> * Local Path Provisioner: 10000
> * Longhorn: 22180

<img src="assets/medium-node-spec/read-performance/random-read-iops-local-path-vs-longhorn.png" alt="drawing" style="width:600px;"/>

**Comment**: 
* Because Longhorn has 3 replicas, it can read from 3 replicas concurrently thus may produce better read performance

> Next, we use Kbench with Longhorn PVCs and scale up the number of Kbench pods to see how random read IOPs is effected when there are more and more IO intensive pods.

Author: Thank you. Fixed

            - containerPort: 8080
              name: metrics
          volumeMounts:
            - name: vol
              mountPath: /volume/
            - name: shared-data
              mountPath: /test-result
      volumes:
        - name: shared-data
          emptyDir: {}
  volumeClaimTemplates:
    - metadata:
        name: vol
      spec:
        accessModes: ["ReadWriteOnce"]
        storageClassName: "longhorn"
        resources:
          requests:
            storage: 20Gi
```
</details>

Some of the important Kbench parameters we would like to call out are:
* `MODE: random-write-iops`: specifies that we are running [random write iops job](https://github.com/longhorn/kbench/blob/main/fio/iops-random-write.fio)
* `SIZE: 15G`: the `fio` test size is 15G, this will avoid cache effect
* PVC size is 20G

> Result:
> * Local Path Provisioner: 9980
> * Longhorn: 1150

<img src="assets/medium-node-spec/write-performance/random-write-iops-local-path-vs-longhorn.png" alt="drawing" style="width:600px;"/>

**Comment**:
* Longhorn write speed is slower than local path volume because of several factors:
  * Because Longhorn has 3 replicas, it has to write to 3 replicas for each IO

>   * By default, Longhorn has revision counter enabled which means that for each write IO, Longhorn has to do an additional write IO to write the revision counter

Reviewer 2: This will need to be revisited and the numbers re-run when the default changes, of course.  But it's right to use the default for v1.6.0.

<summary>StatefulSet workload</summary>
<br>

```yaml
apiVersion: v1
kind: Service
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  ports:
  - port: 80
    name: web
  selector:
    app: nginx
  type: NodePort
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: web
  namespace: default
spec:
  podManagementPolicy: Parallel
  selector:
    matchLabels:
      app: nginx
  serviceName: "nginx"
  replicas: 0
  template:
    metadata:
      labels:
        app: nginx
    spec:
      restartPolicy: Always
      terminationGracePeriodSeconds: 10
      containers:
      - name: nginx
        image: registry.k8s.io/nginx-slim:0.8
        livenessProbe:
          exec:
            command:
              - ls
              - /usr/share/nginx/html/lost+found
          initialDelaySeconds: 5
          periodSeconds: 5
        ports:
        - containerPort: 80
          name: web
        volumeMounts:
        - name: www
          mountPath: /usr/share/nginx/html
  volumeClaimTemplates:
  - metadata:
      name: www
    spec:
      accessModes: [ "ReadWriteOnce" ]
      storageClassName: "longhorn"
      resources:
        requests:
          storage: 1Gi
```

</details>

> Then we scale up the StatefulSet with the rate 30 pods per minutes and monitor the number of pods, CPU and RAM usage of Longhorn system, ECTD, and KubeAPI metrics. The scaling script is:

Reviewer 2: "per minutes" -> "per minute"

Author: Thank you. Fixed

```bash
NAMESPACE="default"; STATEFULSET_NAME="web"; MAX_PODS="<a-specific-number-for-each-test>"; INTERVAL=2; while true; do CURRENT_REPLICAS=$(kubectl get statefulset -n $NAMESPACE $STATEFULSET_NAME -o jsonpath='{.spec.replicas}'); if [ "$CURRENT_REPLICAS" -lt "$MAX_PODS" ]; then NEW_REPLICAS=$((CURRENT_REPLICAS + 1)); echo "Scaling $STATEFULSET_NAME to $NEW_REPLICAS replicas"; kubectl scale --replicas=$NEW_REPLICAS statefulset -n $NAMESPACE $STATEFULSET_NAME; fi; sleep $INTERVAL; done
```
We define the following stopping conditions for the scale test. 
If we hit any of these conditions, we stop the scale test and record the current scaling number.
1. There are more than 10 pods with starting time more than 4 minutes (the duration from pod being created to pod becomes running). 
   The idea is that if pod is taking more than 4 minutes to start, the system is being too slow
1. The per-minute rate of pod become running is smaller than 10. It means that the system should be able to create more than 10 pods per minute. 
   If not, it is being too slow
1. There are more than 3 crashing pods or restarted containers. If there are many crashes, the system is not stable at the scale

> Finally, we will scale down the StatefulSet from the max value to 0 with the rate of 30 pods per minutes. We will measure the scaling down speed.

Reviewer 2: "per minutes" -> "per minute"

Author: Thank you. Fixed

This test design touches the end-to-end life cycle of a pod from pod creation -> CSI flow provision/attach/mount Longhorn volume -> CSI flow unmount/detach Longhorn volume -> pod termination

> > Notes: There are 2 small modification we made in this section comparing to the previous data plan performance testing section:

Reviewer 2: three small modifications.  Or maybe "a few small modifications".

Author: Thank you. Fixed

### 1 master node + 3 worker nodes
#### Result
1. The cluster can handle up to 699 pods with 699 Longhorn volumes
1. Each node can have 233 Longhorn volumes.
2. Each node can have 669 Longhorn replicas.

#### Scaling Up
<img src="./assets/medium-node-spec/control-plane-performance/1-master-3-workers-scale-up.png" alt="drawing"/>

##### Comment
1. This test hits the 1st stopping condition (there are more than 10 pods with starting time more than 4 minutes) at 699 pods (with 699 Longhorn volumes).
1. We can see that the number of running pod is linearly scaled from 0 to 460. Then rate is slow down after that.
1. The component which is consuming the most CPU and RAM are instance manager pods followed by longhorn manager pods
1. API and ETCD are not overloaded

#### Scaling Down

<img src="./assets/medium-node-spec/control-plane-performance/1-master-3-workers-scale-down.png" alt="drawing"/>

> ##### Comment

Reviewer 2: Worth mentioning in the comment that some of the graphs are clearly not meaningful because they reflect starting time and no pods are starting. 

Do you think there would be any reason to look at stopping time? I would think not, since the memory and cpu graphs are so clean.

Author: Thank you. Fixed

Reviewer 2: If it is not too laborious, it would be worth closing the loop on for completeness.

> * At 15 pods, the average random read IOPs per Longhorn volume is 2116. Total random IOPs is 31740

<img src="./assets/medium-node-spec/read-performance/random-read-iops-6-9-12-15.png" alt="drawing" style="width:600px;"/>

Note: the areas with the while background in the graph are transition periods. We can ignore the data in these areas. 

**Comment**:
* From the scaling test so far, we can see that the total random read IOPs of all Longhorn volumes remains relatively the same (around 31500) when the number of Kbench pods increase.
  If we call the average random read IOPs each volume can achieve (x) and the number of volumes (y), they form a reciprocal function: x * y = 31500. 
  Users can use this information to make some prediction for this cluster:
  * The upper bound limit that Longhorn system can achieve in this cluster is 31500 random read IOPs
  * If each workload pod is doing 1000 random IOPs on average, there can be about 31 pods
* When the user keeps scaling up number of pods, eventually this reciprocal relation (x * y = 31500) might no longer hold as the CPU contention and other factors kick in (i.e. x*y will be less and less)

> * The bottleneck in this cluster seems to be the IOPs performance of the EBS volumes on host instead of CPU, memory, or network bandwidth

Reviewer 3: NIT: It's better to re-format this part a little bit. eg
```
##### Scale: 1 Workload pod

**Result**:
- xxxx
- xxxx

<img src=xxxxx>

**Analysis & Conclusion**:
- xxxxxx
- xxxxxx

##### Scale: 3 Workload pods

**Result**:
- xxxx
- xxxx

<img src=xxxxx>

**Analysis & Conclusion**:
- xxxxxx
- xxxxxx

##### Scale: >3 Workload pod

**Result**:
- xxxx
- xxxx

<img src=xxxxx>

**Analysis & Conclusion**:
- xxxxxx
- xxxxxx
```

Author: Thank you for the feedback. 
* I edit the report to replace `**Comment**:` by `**Analysis & Conclusion**:` 
* I do not do this **Result**: because I would like to highlight the result section inside a quote style like:
  > Result:
  > * Local Path Provisioner: 803 microsecond
  > * Longhorn: 2028 microsecond
* I do not do this `##### Scale: 3 Workload pods` because I think there are too many headers in this report already so I don't want to create more tiny headers. Plus not all test cases will require this type of scaling 

WDYT?

Reviewer 3: With the 1st glance at this format, I was a little lost about which case the "result" and "comments" belong to. That's why I prefer to add one more layer of header.

Author: Ok. Adding it. Thank you

Author: Done

            - containerPort: 8080
              name: metrics
          volumeMounts:
            - name: vol
              mountPath: /volume/
            - name: shared-data
              mountPath: /test-result
      volumes:
        - name: shared-data
          emptyDir: {}
  volumeClaimTemplates:
    - metadata:
        name: vol
      spec:
        accessModes: ["ReadWriteOnce"]
        storageClassName: "longhorn"
        resources:
          requests:
            storage: 20Gi
```
</details>

Some of the important Kbench parameters we would like to call out are:
* `MODE: sequential-read-bandwidth`: specifies that we are running [sequential read bandwidth job](https://github.com/longhorn/kbench/blob/main/fio/bandwidth-sequential-read.fio)
* `SIZE: 15G`: the `fio` test size is 15G, this will avoid cache effect
* PVC size is 20G

> Result:
> * Local Path Provisioner: 362 MiB/s
> * Longhorn: 669 MiB/s

<img src="./assets/medium-node-spec/read-performance/sequential-read-bandwidth-local-path-vs-longhorn.png" alt="drawing" style="width:600px;"/>

**Comment**:
* Because Longhorn has 3 replicas, it can read from 3 replicas concurrently thus may produce better read performance

> * Note that the sequential read bandwidth is smaller than the random read bandwidth when testing against a single Longhorn volume of 3 replicas.

Reviewer 3: NIT: We can use a special symbol or something like Italic to mark the To-Be-Investigated issues.

Author: Thank you. Fixed

              value: read-only
            - name: RATE_LIMIT_TYPE
              value: no-rate-limit
          ports:
            - containerPort: 8080
              name: metrics
          volumeMounts:
            - name: vol
              mountPath: /volume/
            - name: shared-data
              mountPath: /test-result
      volumes:
        - name: shared-data
          emptyDir: {}
  volumeClaimTemplates:
    - metadata:
        name: vol
      spec:
        accessModes: ["ReadWriteOnce"]
        storageClassName: "longhorn"
        resources:
          requests:
            storage: 20Gi
```
</details>

Some of the important Kbench parameters we would like to call out are:
* `MODE: random-read-latency`: specifies that we are running [random read latency job](https://github.com/longhorn/kbench/blob/main/fio/latency-random-read.fio)
* `SIZE: 15G`: the `fio` test size is 15G, this will avoid cache effect
* PVC size is 20G

> Result:
> * Local Path Provisioner: 500 microsecond
> * Longhorn: 750 microsecond

<img src="./assets/medium-node-spec/read-performance/random-read-latency-local-path-vs-longhorn.png" alt="drawing" style="width:600px;"/>

**Comment**:

> * Because the IO path of a Longhorn volume is longer than local path provisioner, it is expected that the Longhorn volume has bigger random read latency 

Reviewer 3: NIT: "longer" -> "more complicated"

Author: Thank you. Fixed

Scaling workload from 1 to 3 pods.
> Result:
> * Each Kbench pod is able to achieve 1180 random write IOPs on its Longhorn volume
> * Total random write IOPs can be achieved by all 3 Longhorn volumes is 3540

<img src="assets/medium-node-spec/write-performance/random-write-iops-1-3.png" alt="drawing" style="width:600px;"/>

Scaling workload from 3 to 6, then 6 to 9, then 9 to 12, then 12 to 15
> Result:
> * At 6 pods, the average random write IOPs per Longhorn volume is 852. Total random write IOPs is 5112
> * At 9 pods, the average random write IOPs per Longhorn volume is 576. Total random write IOPs is 5184
> * At 12 pods, the average random write IOPs per Longhorn volume is 436. Total random write IOPs is 5232
> * At 15 pods, the average random write IOPs per Longhorn volume is 350. Total random write IOPs is 5250

<img src="./assets/medium-node-spec/write-performance/random-write-iops-6-9-12-15.png" alt="drawing" style="width:600px;"/>

**Comment**:
* From the scaling test so far, we can see that the total random write IOPs of all Longhorn volumes remain relatively the same around 5250 when the number of Kbench pods increase

> * Since each EBS volume on the host is provisioned with 10000 IOPs. It looks like Longhorn system is able to reach half of the IOPs capacity of the host EBS volumes

Reviewer 3: NIT: You can explicitly mention why it's a half of EBS capacity. (Revision counter)

Author: Thank you. Fixed

> Result:
> * Each Kbench pod is able to achieve 2350 sequential write IOPs on its Longhorn volume
> * Total sequential write IOPs can be achieved by all 3 Longhorn volumes is 7050

<img src="assets/medium-node-spec/write-performance/sequential-write-iops-1-3.png" alt="drawing" style="width:600px;"/>

Scaling workload from 3 to 6, then 6 to 9, then 9 to 12, then 12 to 15
> Result:
> * At 6 pods, the average sequential write IOPs per Longhorn volume is 1745. Total sequential write IOPs is 10470
> * At 9 pods, the average sequential write IOPs per Longhorn volume is 1165. Total sequential write IOPs is 10485
> * At 12 pods, the average sequential write IOPs per Longhorn volume is 871. Total sequential write IOPs is 10452
> * At 15 pods, the average sequential write IOPs per Longhorn volume is 695. Total sequential write IOPs is 10425

<img src="./assets/medium-node-spec/write-performance/sequential-write-iops-6-9-12-15.png" alt="drawing" style="width:600px;"/>

**Comment**:
* From the scaling test so far, we can see that the total sequential write IOPs of all Longhorn volumes remain relatively the same around 10400 when the number of Kbench pods increase

> * Since each EBS volume on the host is provisioned with 10000 IOPs. It looks like Longhorn system is able to reach the IOPs capacity of the host EBS volumes

Reviewer 3: We can analyze why here Longhorn volume sequential write could reach the IOPs capacity of the host EBS volumes:
Due to the IO merge, Longhorn does not need to update the revision counter frequently. Then the degradation caused by the revision counter is greatly reduced.

Author: Thank you. Fixed

   1. For volume with full data pattern (A.K.A every snapshot is a full), the maximum volume size would be calculated as `30Ti/x` where `x` is maximum number of snapshots setting of the volume (including volume-head)
   1. For volume with data pattern 4k_data-4k_hole-4k_data-4k_hole-..., the maximum volume size would be calculated as `772Gi/x` where `x` is maximum number of snapshots setting of the volume (including volume-head). This is the worst case
1. What is the maximum IOPs of a Longhorn volume?  
   1. random-read-iops: 22180
   1. sequential-read-iops: 39772
   1. random-write-iops: 1150
   1. sequential-write-iops: 2340
1. What is the maximum bandwidth of a Longhorn volume?
    1. random-read-bandwidth: 874 MiB/s
    1. sequential-read-bandwidth: 669 MiB/s
    1. random-write-bandwidth: 159 MiB/s
    1. sequential-write-bandwidth: 291 MiB/s
1. What is the minimum latency of Longhorn volume?
    1. random-read-latency: 750 microseconds
    1. sequential-read-latency: 740 microseconds
    1. random-write-latency:  2040 microseconds
    1. sequential-write-latency: 2028 microseconds
1. What is the maximum IOPs of all Longhorn volumes with 3 worker nodes?

>     1. random-read-iops: ~31700

Reviewer 3: NIT: `random-read-iops: ~31700 (The sum of all worker node EBS volumes' capability)`

Same below.

Author: I would prefer to let the user read the details in the corresponding test section to understand that it is `The sum of all worker node EBS volumes' capability`. I don't want to add too many details in this FAQs section because: if we add `random-read-iops: ~31700 (The sum of all worker node EBS volumes' capability)` then it doesn't make sense that the below line is `sequential-read-iops: ~62500`. We would have to further more explaining that due to IO merging. Then something for `random-write-iops: ~5250` vs `sequential-write-iops: ~10400`, we would need to explain more. So I think it is better to leave the detailed explanation inside each test section. WDYT?

Reviewer 3: OK. I am fine with either. 
To be honest, I don't think normal users will try to remember all the details when reading this report.

              mountPath: /test-result
      volumes:
        - name: shared-data
          emptyDir: {}
  volumeClaimTemplates:
    - metadata:
        name: vol
      spec:
        accessModes: ["ReadWriteOnce"]
        storageClassName: "longhorn"
        resources:
          requests:
            storage: 20Gi
```
</details>

Some of the important Kbench parameters we would like to call out are:
* `MODE: random-read-iops`: specifies that we are running [random read iops job](https://github.com/longhorn/kbench/blob/main/fio/bandwidth-random-read.fio)
* `SIZE: 15G`: the `fio` test size is 15G, this will avoid cache effect
* PVC size is 20G

**Result**: 
> * Local Path Provisioner: 10000
> * Longhorn: 22180

<img src="assets/medium-node-spec/read-performance/random-read-iops-local-path-vs-longhorn.png" alt="drawing" style="width:600px;"/>

**Analysis & Conclusion**: 
* Because Longhorn has 3 replicas, it can read from 3 replicas concurrently thus may produce better read performance

Next, we use Kbench with Longhorn PVCs and scale up the number of Kbench pods to see how random read IOPs is affected when there are more and more IO intensive pods.

##### 3 Workload Pods Scale

> Scaling workload from 1 to 3 pods.

Reviewer 3: NIT: Missing an empty line

Author: Thank you. Fixed

   1. [Random Write Latency - Stress Tests](#random-write-latency---stress-tests)
   1. [Sequential Write Latency - Stress Tests](#sequential-write-latency---stress-tests)
1. [Longhorn Control Plane Performance](#longhorn-control-plane-performance)
1. [Volume Maximum Size](#volume-maximum-size)

> 1. [FQA](#faq)

Reviewer 2: FAQ

Author: Thank you. Fixed
